modin/engines/ray/generic/io.py

```diff
@@ -144,24 +147,48 @@ def read_parquet(cls, path, engine, columns, **kwargs):
             columns[i : i + column_splits]
             for i in range(0, len(columns), column_splits)
         ]
-        # Each item in this list will be a list of columns of original df
-        # partitioned to smaller pieces along rows.
-        # We need to transpose the oids array to fit our schema.
-        blk_partitions = np.array(
-            [
-                cls.read_parquet_remote_task._remote(
-                    args=(path, cols, num_splits, kwargs),
-                    num_return_vals=num_splits + 1,
-                )
-                for cols in col_partitions
-            ]
-        ).T
+     
+        pf = ParquetFile(path)
+        column_names = [
+                  name
+                  for name in pf.metadata.schema.names
+                  if not PQ_INDEX_REGEX.match(name)
+              ]
+
+        num_row_groups = pf.metadata.num_row_groups 
+     
+
+        blocks_step1 =  [ 
```

Comment:
We may need to rethink the idea that each `row_group` + `column` combination returns the `index` or the length of the `index`. This is creating multiple copies.

It would be good to try to get this to work with 2D list comprehension. Example:

```python
blocks = [
    [
        cls.read_parquet_remote_task._remote(
            args=(path, col_partitions[i], num_splits=1, row_group=j, kwargs),
            num_return_vals=1 if j != num_row_groups - 1 else 2
        ) for i in range(len(column_splits))
    ] 
    for j in range(num_row_groups)
]
```

Of course we would have to add the logic for handling the last column group, but we should take advantage of the `col_partitions` variable created above to avoid over-splitting the data.

Reply:
When you say above, "this creates multiple copies",  can you describe for me in more details what object/data exactly you are referring to that is getting copied multiple times?  Do you mean the redundant length numbers ( which are indeed the same for all columns), though in the case here where it is just and integer, not actually and index, seems not too expensive.

Reply:
Right, the index (which is an integer in the common case). This is expensive because there are fixed overheads to each `ray.put`. We want to avoid putting data in the object store that is redundant/never needed.

Reply:
Let's try that. We can always iterate.